File: py-rosalind/dna.py
class NucleotideCount:
    def __init__(self, a, c, g, t):
        self.A = a
        self.C = c
        self.G = g
        self.T = t

    def __str__(self):
        return f'{self.A} {self.C} {self.G} {self.T}'

    def total(self):
        return self.A + self.C + self.G + self.T

def dna(stream):
    A = C = T = G = 0
    for line in stream:
        for c in line:
            if c == 'A':
                A += 1
            elif c == 'C':
                C += 1
            elif c == 'G':
                G += 1
            elif c == 'T':
                T += 1

    return NucleotideCount(A, C, G, T)

File: rust-rosalind/src/dna.rs
use std::io::{BufRead, BufReader};
use std::fmt;

pub struct NucleotideCount {
    pub a: u32,
    pub c: u32,
    pub g: u32,
    pub t: u32
}

impl NucleotideCount {
    pub fn total(&self) -> u32 {
        self.a + self.c + self.g + self.t
    }
}

impl fmt::Display for NucleotideCount {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} {} {} {}", self.a, self.c, self.g, self.t)
    }
}

pub fn dna_count(s: &String) -> NucleotideCount {
    let (mut a, mut c, mut g, mut t) = (0, 0, 0, 0);
    for letter in s.chars() {
        if letter == 'A' { a += 1; }
        else if letter == 'C' { c += 1; }
        else if letter == 'G' { g += 1; }
        else if letter == 'T' { t += 1; }
    }

    NucleotideCount { a: a, c: c, g: g, t: t }
}

pub fn dna(reader: BufReader<std::fs::File>) -> NucleotideCount {
    let (mut a, mut c, mut g, mut t) = (0, 0, 0, 0);
    for line in reader.lines() {
        let line = line.expect("Could not read next line");
        for letter in line.chars() {
            if letter == 'A' { a += 1; }
            else if letter == 'C' { c += 1; }
            else if letter == 'G' { g += 1; }
            else if letter == 'T' { t += 1; }
        }
    }

    NucleotideCount { a: a, c: c, g: g, t: t }
}